credentials/apps/badges/models.py

```diff
@@ -83,6 +83,35 @@ def save(self, *args, **kwargs):
     @classmethod
     def by_uuid(cls, template_uuid):
         return cls.objects.filter(uuid=template_uuid, origin=cls.ORIGIN).first()
+    
+    def user_progress(self, username: str) -> float:
+        """
+        Calculate user progress for badge template.
+        """
+        requirements_count = BadgeRequirement.objects.filter(template=self).count()
+        if requirements_count == 0:
+            raise ValueError("No requirements found for badge template")
+
+        fulfilled_requirements_count = Fulfillment.objects.filter(progress__username=username, requirement__template=self).count()
+        return fulfilled_requirements_count / requirements_count
+    
+    def user_completion(self, username: str) -> bool:
+        """
+        Check if user completed badge template.
+        """
+        return self.user_progress(username) == 1.0
+    
+    @property
+    def ratio(self) -> float:
```

Comment:
This property is expected to be on the `BadgeProgress`, so it is possible to get the exact badge template progress value on a specific `BadgeProgress` record (not a generic ratio on a BadgeTemplate).

So, the `.ratio` value must mirror the `.user_progress(username)` but from the progress instance side.

Reply:
Sorry, if this wasn't that clear in the description 😞